Implement RegExp dotAll flag (/s)

rhuanjl opened the pull request:
This PR implements the es2018 RegExp /s dotAll flag.

See proposal spec here: https://tc39.github.io/proposal-regexp-dotall-flag/

Notes:
1. I've given this the flag ES9RegExDotAll but I've made it default to enabled
2. When run with the ES6RegExPrototypeProperties this passes all relevant test262 tests, without that flag several test262 tests fail (hopefully that can be enabled soon)

fixes: #2787 

cc: @mathiasbynens

sethbrenith commented on test/es6/regexflags.js:
```diff
@@ -57,28 +58,37 @@ function generatePrototypeFlagsTests() {
                 r.sticky = true; // no-op
+                r.dotAll = true; // no-op
                 assert.areEqual(expectedUnicode, r.unicode, r + ": unicode");
                 assert.areEqual(expectedSticky, r.sticky, r + ": sticky");
```
Looks like you could add a third assertion here for expectedDotAll

rhuanjl replied:
Oops I thought I had I'll update that.

sethbrenith commented on lib/Runtime/Library/JavascriptRegularExpression.cpp:
```diff
@@ -1516,6 +1561,8 @@ using namespace Js;
         case PropertyIds::options:
             IS_WRITABLE(!scriptConfig->IsES6RegExPrototypePropertiesEnabled())
+        case PropertyIds::dotAll:
+            IS_WRITABLE(!scriptConfig->IsES9RegExDotAllEnabled())
```
Why does this one not follow the pattern of the others? In every other method the corresponding case looks like `(scriptConfig->IsES9RegExDotAllEnabled() && !scriptConfig->IsES6RegExPrototypePropertiesEnabled())`.

kfarnung commented:
@dilijev Can you take a look at this change?

dilijev commented on lib/Parser/RegexRuntime.cpp:
```diff
@@ -5756,6 +5756,7 @@ namespace UnifiedRegex
         if ((flags & GlobalRegexFlag) != 0) w->Print(_u("global "));
         if ((flags & MultilineRegexFlag) != 0) w->Print(_u("multiline "));
         if ((flags & IgnoreCaseRegexFlag) != 0) w->Print(_u("ignorecase"));
+        if ((flags & DotAllRegexFlag) != 0) w->Print(_u("dotAll"));
```
>if ((flags & DotAllRegexFlag) != 0) w->Print(_u("dotAll")); [](start = 8, length = 59)

nit: all of these strings should have a space at the end (like `"global "` above) so that they are spaced out correctly when printed

dilijev commented on lib/Parser/StandardChars.cpp:
```diff
@@ -261,6 +261,11 @@ END {
 
+    void StandardChars<char16>::SetFullSet(ArenaAllocator* setAllocator, CharSet<Char> &set)
+    {
+        set.SetRange(allocator, MinChar, MaxChar);
```
>set.SetRange [](start = 8, length = 12)

This is fine as-is, but for memory usage reasons (and some runtime efficiency), I'd prefer if you could make the set the equivalent of `[^<none>]` instead of the equivalent of `[\x00-\u{1FFFFF}]` as you've done here. You should be able to use `SetNotRanges(0, <empty string>)` to accomplish this, but if there's some assertions or preconditions there that you're not using empty string, you might need a new method.

rhuanjl replied:
SetNotRanges with 0 sorted pairs currently literally does what I've done here:
https://github.com/Microsoft/ChakraCore/blob/bffb8850cd491ba148b98c4ee9f322312d374c56/lib/Parser/CharSet.cpp#L1137-L1150

All the NotRanges seem to be handled in that way.

dilijev replied:
Ah, that is surprising to me since we have a concept of a negated set which is more memory efficient. Possibly that was a later addition than the sets themselves. I'll look into this at a later time; won't block the PR on it.

rhuanjl replied:
I'll change dotAll one to use SetNotRanges so if that method is updated to use negated sets this will benefit from the change.

dilijev replied:
Sounds good!

dilijev commented on lib/Runtime/Library/JavascriptRegularExpression.cpp:
```diff
@@ -1281,6 +1312,8 @@ using namespace Js;
         case PropertyIds::options:
             SET_PROPERTY(!scriptConfig->IsES6RegExPrototypePropertiesEnabled());
+        case PropertyIds::dotAll:
+            SET_PROPERTY(scriptConfig->IsES9RegExDotAllEnabled() && !scriptConfig->IsES6RegExPrototypePropertiesEnabled());
```
>ES9 [](start = 41, length = 3)

does this ES2018 feature check make sense with the ES6 check?

dilijev replied:
Yeah I guess this would have to be blocked on the same feature.

---
In reply to: [210786581](https://github.com/Microsoft/ChakraCore/pull/5592#discussion_r210786581) [](ancestors = 210786581)

rhuanjl replied:
It's not blocked on the ES6 feature check it's rather that all of these properties behave differently without the ES6 Prototype properties enabled.

With ES2018RegExDotAllEnabled and ES6RegExPrototypePropertiesEnabled the prototype has a dotAll getter only property which returns if the RegEx it's called on has the dotAll flag.

If dotAll is enabled but the Prototype properties are not enabled it's a property directly on the RegExp object which returns true (as opposed to a getter from the prototype)

dilijev replied:
Right, I had a look and found the same. Thanks for clarifying.

rhuanjl commented:
@dilijev thanks for the review, I've addressed most of your points and updated to latest. I've held off on squashing for now in case you wish to check the new changes separately.

But I'm afraid I don't know how to address the SetRange point - see my comment above.

rhuanjl commented:
Thanks @dilijev 
Any recent pain on ByteCode regens has likely been partly my fault considering I've got this one +  Object.fromEntries open and I also recently had merged Array.flat/flatMap.

Hopefully this is good to merge now (Note I've left the Bytecode regen as a separate commit so it can be easily dropped if there's another merge conflict - obviously either this or Object.fromEntries will have to have another regen...)
